libmusig2.c

```diff
@@ -1,282 +1,359 @@
-        crypto_core_ristretto255_scalar_add(response, response, temp_scalar);
-    }
-}
+void SIG_SecondRound(const secp256k1_context* ctx, PARAM_ROUND2 paramRound2, SIGNER signer, unsigned char* SIGNATURE, unsigned char* msg){
 
-void compute_challenge(unsigned char *challenge,
-                      const unsigned char *aggr_pks,
-                      const unsigned char *announcement,
-                      const unsigned char *m,
-                      unsigned long long mlen
-        ) {
-    // we use MAX (64) instead of normal (32) to get almost uniformity.
-    unsigned char hash[64];
-    // before hashing we need to change the encoding of the points, to make
-    // sure that the verifier gets the same challenge
-    unsigned char pk_ed25519[crypto_core_ed25519_BYTES];
-    unsigned char announcement_ed25519[crypto_core_ed25519_BYTES];
-
-    // Now we need to get the torsion-free representative of the ristretto points in edwards form, to ensure
-    // that the verification equation will validate (the verifier will handle torsion-free elements).
-
-    // Now we need to get the torsion-free representative of the ristretto points in edwards form, to ensure
-    // that the verification equation will validate.
-    if (map_ristretto_prime_subgroup(pk_ed25519, aggr_pks) == -1 ||
-            map_ristretto_prime_subgroup(announcement_ed25519, announcement) == -1) {
-        printf("conversion went wrong");
+    int i,j;
+    unsigned char* rb = (unsigned char*) malloc(SK_SIZE);
+    unsigned char* sum_rb = (unsigned char*) malloc(SK_SIZE);
+    unsigned char a[32];
+    unsigned char c[32];
+
+
+    /* Compute the key aggregate coefficient (a) of signer. */
+    if(!secp256k1_tagged_sha256(ctx, a, signer->PK, PK_SIZE, paramRound2->L, PK_SIZE*N)){
+        printf("Failed to generate key aggregation coefficient!\n");
+        return;
     }
 
-    crypto_hash_sha512_state state;
-    crypto_hash_sha512_init(&state);
-    crypto_hash_sha512_update(&state, announcement_ed25519, 32);
-    crypto_hash_sha512_update(&state, pk_ed25519, 32);
-    crypto_hash_sha512_update(&state, m, mlen);
+    /* Compute c. */
+    CALC_C(ctx,paramRound2->X_,paramRound2->R_state,msg,c);
+
+    /* Copy the signer's secret key into signature. */
+    for (i = 0; i < SK_SIZE; i++)
+        SIGNATURE[i] = signer->sk[i];
 
-    crypto_hash_sha512_final(&state, hash);
+    /* Calculate a*c*sk. */
+    if(!(secp256k1_ec_seckey_tweak_mul(ctx,SIGNATURE,a) &&
+            secp256k1_ec_seckey_tweak_mul(ctx,SIGNATURE,c)))
+    {
+        printf("Failed to calculate signature!\n");
+        return;
+    }
 
-    crypto_core_ristretto255_scalar_reduce(challenge, hash);
+    for (j = 0; j < V; j++){
+        if(j==0){
+            /* If j is 0, then the first entry of sum list will be signers r[j] itself.
+             * So, copy the content of r[j] into sum_rb. */
+            for (i = 0; i < SK_SIZE; i++)
+                sum_rb[i] = signer->r[j][i];
```

Comment:
as a safety measure we should zero out the memory of rs after using them (to avoid using them twice, specially if we allow for batches of committed values.